File: src/main/java/com/ullgren/modern/simple3d/StarField.java
package com.ullgren.modern.simple3d;

import java.awt.Color;
import java.awt.Graphics;
import java.util.Random;

/**
 * A static starfield background of {@value #STAR_COUNT} stars.
 * <p>
 * Stars are stored as normalised (0–1) coordinates so they scale correctly when the canvas
 * is resized. Each star has an independently chosen size and brightness.
 * The layout is deterministic (seed 42).
 */
public class StarField {

  private static final int STAR_COUNT = 200;

  private final float[] starX      = new float[STAR_COUNT];
  private final float[] starY      = new float[STAR_COUNT];
  private final float[] starRadius = new float[STAR_COUNT];
  private final float[] starAlpha  = new float[STAR_COUNT];

  public StarField() {
    Random rng = new Random(42);
    for (int i = 0; i < STAR_COUNT; i++) {
      starX[i]      = rng.nextFloat();
      starY[i]      = rng.nextFloat();
      // Most stars are tiny (radius 0.5–1.5 px); a few are larger
      starRadius[i] = 0.5f + rng.nextFloat() * rng.nextFloat() * 2.0f;
      starAlpha[i]  = 0.4f + rng.nextFloat() * 0.6f;
    }
  }

  /**
   * Draws the star field onto {@code g} for a canvas of size {@code w} × {@code h}.
   */
  public void draw(Graphics g, int w, int h) {
    for (int i = 0; i < STAR_COUNT; i++) {
      int alpha = (int) (starAlpha[i] * 255);
      g.setColor(new Color(255, 255, 255, alpha));
      int x = (int) (starX[i] * w);
      int y = (int) (starY[i] * h);
      int d = Math.max(1, Math.round(starRadius[i] * 2));
      if (d <= 1) {
        g.drawLine(x, y, x, y);
      } else {
        g.fillOval(x, y, d, d);
      }
    }
  }
}

File: src/web/src/render/StarField.ts
/**
 * Static starfield background of 200 stars.
 *
 * Stars use normalised (0–1) coordinates so they scale with the canvas.
 * Layout is deterministic via a port of Java's java.util.Random(seed=42),
 * producing identical star positions to the Java version.
 */
export class StarField {
  private static readonly COUNT = 200;

  private readonly starX:      Float32Array;
  private readonly starY:      Float32Array;
  private readonly starRadius: Float32Array;
  private readonly starAlpha:  Float32Array;

  constructor() {
    const n = StarField.COUNT;
    this.starX      = new Float32Array(n);
    this.starY      = new Float32Array(n);
    this.starRadius = new Float32Array(n);
    this.starAlpha  = new Float32Array(n);

    const rng = new JavaRandom(42);
    for (let i = 0; i < n; i++) {
      this.starX[i]      = rng.nextFloat();
      this.starY[i]      = rng.nextFloat();
      this.starRadius[i] = 0.5 + rng.nextFloat() * rng.nextFloat() * 2.0;
      this.starAlpha[i]  = 0.4 + rng.nextFloat() * 0.6;
    }
  }

  draw(ctx: CanvasRenderingContext2D, w: number, h: number): void {
    for (let i = 0; i < StarField.COUNT; i++) {
      const alpha = this.starAlpha[i];
      const x = this.starX[i]  * w;
      const y = this.starY[i]  * h;
      const d = Math.max(1, Math.round(this.starRadius[i] * 2));

      ctx.fillStyle = `rgba(255,255,255,${alpha.toFixed(3)})`;
      if (d <= 1) {
        ctx.fillRect(Math.round(x), Math.round(y), 1, 1);
      } else {
        ctx.beginPath();
        ctx.arc(x, y, d / 2, 0, Math.PI * 2);
        ctx.fill();
      }
    }
  }
}

/**
 * Port of java.util.Random's linear congruential generator.
 * Uses BigInt internally for exact 48-bit arithmetic.
 */
class JavaRandom {
  private seed: bigint;

  constructor(seed: number) {
    this.seed = (BigInt(seed) ^ 0x5DEECE66Dn) & ((1n << 48n) - 1n);
  }

  private next(bits: number): number {
    this.seed = (this.seed * 0x5DEECE66Dn + 0xBn) & ((1n << 48n) - 1n);
    return Number(this.seed >> BigInt(48 - bits));
  }

  nextFloat(): number {
    return this.next(24) / (1 << 24);
  }
}
